pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go

```diff
@@ -0,0 +1,316 @@
+				Expect(err).ToNot(HaveOccurred())
+				Expect(cert).ToNot(BeNil())
+			})
+
+			It("should return an error - kubeconfig secret does not exist", func() {
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				_, err := getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).To(HaveOccurred())
+			})
+
+			It("should return an error - secret does not contain a kubeconfig", func() {
+				// mock existing secret with garden kubeconfig
+				kubeconfigSecret := &corev1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      gardenClientConnection.KubeconfigSecret.Name,
+						Namespace: gardenClientConnection.KubeconfigSecret.Namespace,
+					},
+				}
+				err := k8sClient.Create(context.TODO(), kubeconfigSecret)
+				Expect(err).ToNot(HaveOccurred())
+
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				_, err = getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).To(HaveOccurred())
+			})
+
+			It("should return an error - kubeconfig client cert is invalid", func() {
+				// mock existing secret with garden kubeconfig
+				updateKubeconfigSecret(k8sClient, gardenClientConnection, gardenKubeconfigWithInValidClientCert)
+
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				_, err := getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).To(HaveOccurred())
+			})
+		})
+
+		Describe("#waitForCertificateRotation", func() {
+			var fakeClientMap clientmap.ClientMap
+
+			BeforeEach(func() {
+				k8sGardenClient.EXPECT().DirectClient().Return(k8sClient).AnyTimes()
+
+				validity := 20 * time.Second
+				cert := generateCertificate(validity)
+
+				testKubeconfig := fmt.Sprintf(baseKubeconfig, utils.EncodeBase64(cert.CertificatePEM), utils.EncodeBase64(cert.PrivateKeyPEM))
+				updateKubeconfigSecret(k8sClient, gardenClientConnection, testKubeconfig)
+
+				fakeClientMap = fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+			})
+
+			It("should not return an error", func() {
+				err := waitForCertificateRotation(context.TODO(), log, time.Now, gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("should return an error - the certificates expiration date changed", func() {
+				// simulate changing the certificate while waiting for the certificate rotation deadline
+				go func() {
+					timer := time.NewTimer(10 * time.Second)
+					defer timer.Stop()
+
+					// block for 10 seconds
+					<-timer.C
```

Comment:
Please avoid such time-sensitive unit tests and rather mock this stuff out, as we had many flakes caused by tests like this.
Ref #2614, #2609

Reply:
Using a mock client now - good suggestion.